feat: Update sBTC best practices (#1709)
stacks-network/docs: guides-and-tutorials/sbtc/best-practices-for-running-an-sbtc-signer.md

- Periodically verify the integrity of backups, as instructed below.

### Verifying integrity of PostgreSQL DB backups

To verify the integrity of a backup, first import it into a fresh PostgreSQL
instance (the database is enough, no need to spin up a Stacks / Bitcoin node or
the sBTC signer).

Then, perform the following query:

```sql
signer=> SELECT aggregate_key FROM sbtc_signer.dkg_shares WHERE
dkg_shares_status = 'verified' ORDER BY created_at DESC;
```

It will return results as follows (your mileage might vary depending on the
history of your sBTC signer, the following is provided for illustration purposes
only):

```sql

                            aggregate_key
----------------------------------------------------------------------
 \x03d8c4344861fc7590fd812c24884a3bfd9374d8ba865a787ff53c9060020aa967
 \x03f898f8a6ddb86dd4608dd168355ec6135fe2839222240c01942e8e7e50dd4c89
(2 rows)
```

Now, ensure that the most recent `aggregate_key` (the first one) corresponds to
the one returned by a read-only call to the
`SM3VDXK3WZZSA84XXFKAFAF15NNZX32CTSG82JFQ4/sbtc-registry/get-current-aggregate-pubkey`
smart contract method:

```bash

> curl 'https://api.hiro.so/v2/contracts/call-read/SM3VDXK3WZZSA84XXFKAFAF15NNZX32CTSG82JFQ4/sbtc-registry/get-current-aggregate-pubkey' \
>            -H 'content-type: application/json' --data-raw '{"sender":"SM3VDXK3WZZSA84XXFKAFAF15NNZX32CTSG82JFQ4","arguments":[]}'

Reviewer: ```suggestion
curl -s 'https://api.hiro.so/v2/contracts/call-read/SM3VDXK3WZZSA84XXFKAFAF15NNZX32CTSG82JFQ4/sbtc-registry/get-current-aggregate-pubkey' \
           -H 'content-type: application/json' --data-raw '{"sender":"SM3VDXK3WZZSA84XXFKAFAF15NNZX32CTSG82JFQ4","arguments":[]}' | jq .result
```

> {"okay":true,"result":"0x020000002103d8c4344861fc7590fd812c24884a3bfd9374d8ba865a787ff53c9060020aa967"}⏎

Reviewer: ```suggestion
"0x020000002103d8c4344861fc7590fd812c24884a3bfd9374d8ba865a787ff53c9060020aa967"⏎
```

```

You can discard the prefix `0x02000000210` (which is how Clarity encodes
values). The suffix
`3d8c4344861fc7590fd812c24884a3bfd9374d8ba865a787ff53c9060020aa967` matches the

> last row of the PostgreSQL query above (excluding `x0` which indicates hex

Reviewer: ```suggestion
first row of the PostgreSQL query above (excluding `\x0` which indicates hex
```

encoding).

## Setup proper access control

- Require hardware 2FA keys for access control (e.g., by using Yubikey) to
  connect through SSH, to authenticate to AWS, and for every other relevant
  action.
- Follow the principle of _least privilege_: if you don’t need access, you don’t
  get access; if you get access, it expires after the action is taken.

> - _Optional, but strongly recommended_: Implement a "_4-eyes_" process to access

Reviewer: ```suggestion
- _Optional, but strongly recommended_: Implement a "_4-eyes_" process (require that any activity by an individual must be controlled - reviewed, double checked - by a second individual) to access
```

  critical resources (e.g., deploy a new version of the sBTC signer).

## Maintain a strict firewall configuration

> - Allow connections to your signer `listen_on` address (used for P2P

Reviewer: ```suggestion
- Allow connections to your sBTC signer `listen_on` address (used for P2P
```

  communication).

> - Do not expose any non-essential service to the internet: use a `DEFAULT DENY` policy with explicit `ALLOW`s for necessary network traffic (such as signer p2p and SSH).

Reviewer: ```suggestion
- Do not expose any non-essential service to the internet: use a `DEFAULT DENY` policy with explicit `ALLOW`s for necessary network traffic (such as sBTC signer p2p and SSH).
```

values). The suffix
`3d8c4344861fc7590fd812c24884a3bfd9374d8ba865a787ff53c9060020aa967` matches the
last row of the PostgreSQL query above (excluding `x0` which indicates hex
encoding).

## Setup proper access control

- Require hardware 2FA keys for access control (e.g., by using Yubikey) to
  connect through SSH, to authenticate to AWS, and for every other relevant
  action.
- Follow the principle of _least privilege_: if you don’t need access, you don’t
  get access; if you get access, it expires after the action is taken.
- _Optional, but strongly recommended_: Implement a "_4-eyes_" process to access
  critical resources (e.g., deploy a new version of the sBTC signer).

## Maintain a strict firewall configuration

- Allow connections to your signer `listen_on` address (used for P2P
  communication).
- Do not expose any non-essential service to the internet: use a `DEFAULT DENY` policy with explicit `ALLOW`s for necessary network traffic (such as signer p2p and SSH).

## Maintain a robust secrets management program

- Ensure all relevant secrets are safely managed and rotated (where possible),
  e.g. if someone leaves the team.

## Monitor and observe your sBTC Signer

> - Retain at least 7 day of logs for both the sBTC Signer, the Stacks node, and

Reviewer 2: 30 and 90 are perhaps more standard recommendations.

  e.g. if someone leaves the team.

## Monitor and observe your sBTC Signer

- Retain at least 1 day of logs for both the sBTC Signer, the Stacks node, and
- Retain at least 7 day of logs for both the sBTC Signer, the Stacks node, and
  the Bitcoin node.
- The sBTC signer can optionally expose Prometheus metrics (see
  `prometheus_exporter_endpoint` configuration option).
  - You can use them to monitor its health ([this guide shows how to configure
    Alloy to collect metrics on Grafana
    cloud](../running-a-signer/how-to-monitor-signer.md)).

## Downstream components
## Provision dedicated downstream components

- Run a _dedicated_ Bitcoin node and Stacks node for your sBTC Signer.
  - Ensure the nodes are provisioned with the minimum hardware requirements
    described [here][0].
  - Nodes should be _exclusively dedicated_ to serve the Signer. Avoid
  - Nodes should be _exclusively dedicated_ to serve the sBTC Signer. Avoid
    re-using them to serve other clients as that may negatively affect
    performance (no _mock-signing_, no _Stacks API nodes_).

## Redundancy in operations

- Ensure that multiple, trusted users can manage and maintain your sBTC Signer instance.

> - Where feasible, users should span different timezones.

Author: Thanks, done!

> - Ensure that multiple, trusted users can manage and maintain your sBTC Signer instance.

Reviewer 2: more examples of "users" and not "administrators"

> - Document your operations procedures and ensure that relevant personnel have access to them.

Reviewer 2: Yes, the development SOP (Standard Operating Procedures) are a common way to not only level up but cross train.

The following best practices suggest how to create a resilient setup for running
your sBTC Signer.

## Protect your private key and have a cold-storage backup

- Prevent unauthorised access to the sBTC Signer private key.
- Keep an offline, secure backup of your sBTC Signer private key (e.g., hardware
  security modules or encrypted storage devices).

## Backup your sBTC Signer PostgreSQL DB

- Perform daily backups of the sBTC Signer PostgreSQL DB.
- Periodically verify the integrity of backups (e.g. by importing them into a

>   fresh PostgreSQL instance).

Reviewer 3: Should we expand a bit on the `verify`? I think we don't want people restoring backup on prod running instances, but not even running parallel prod instances with the same config and old db snapshot (connected to the same signer network). OTOH, it may not be clear what to check after an import (eg, maybe it's a schema only backup, so it doesn't fail but is empty)

Author: Heard, should we tell them to spin up an empty PSQL DB, restore the DB and count the aggregate shares?

Reviewer 3: Yeah given that that one is the table we are mostly interested in, we could ask them to query for the latest share and check its status and aggregate key. For the aggregate key we would want that to match the current one in the smart contract, eg fetching it via:
```
curl 'https://api.hiro.so/v2/contracts/call-read/SM3VDXK3WZZSA84XXFKAFAF15NNZX32CTSG82JFQ4/sbtc-registry/get-current-aggregate-pubkey' -H 'content-type: application/json' --data-raw '{"sender":"SM3VDXK3WZZSA84XXFKAFAF15NNZX32CTSG82JFQ4","arguments":[]}'
```
unless there was a recent rotate key, then it would not match.

A possible issue with the row count is that now it may not be informative, as if we fail DKG verification we would retry, so there may be more rows than expected. But I guess we could just ask them to ensure it's non decreasing maybe?

Author: Done, PTAL?

Reviewer 3: > As of 2025-02-27, the query returns the following

This may not match for people that deleted something, or missed one of the last DKG rounds; it also doesn't differentiate between verified and failed rounds, so I would specify that YMMV in the comment.

I would explicitly add to the query the filter for verified only, and explicit sort by created at.

Author: Thanks, done! PTA(nother)L

Reviewer 3: (I can't resolve this for some reason, feel free to)

## Setup proper access control

- Require hardware 2FA keys for access control (e.g., by using Yubikey) to
  connect through SSH, to authenticate to AWS, and for every other relevant
  action.
- Follow the principle of _least privilege_: if you don’t need access, you don’t
  get access; if you get access, it expires after the action is taken.
- _Optional, but strongly recommended_: Implement a "_4-eyes_" process to access
  critical resources (e.g., deploy a new version of the sBTC signer).

## Maintain a strict firewall configuration

- Allow connections to your signer `listen_on` address (used for P2P
  communication).

> - Do not expose any non-essential service to the internet: deny _all other_

Reviewer 4: Perhaps something more prescriptive in this case:

```suggestion
- Do not expose any non-essential service to the internet: use a `DEFAULT DENY` policy with explicit `ALLOW`s for necessary network traffic (such as signer p2p and SSH).
```

  traffic to _every other port and service_, unless required, e.g. for SSH.

## Maintain a robust secrets management program

- Ensure all relevant secrets are safely managed and rotated (where possible),
  e.g. if someone leaves the team.

## Monitor and observe your sBTC Signer

- Retain at least 7 day of logs for both the sBTC Signer, the Stacks node, and
  the Bitcoin node.
- The sBTC signer can optionally expose Prometheus metrics (see
  `prometheus_exporter_endpoint` configuration option).
  - You can use them to monitor its health ([this guide shows how to configure
    Alloy to collect metrics on Grafana
    cloud](../running-a-signer/how-to-monitor-signer.md)).

## Provision dedicated downstream components

- Run a _dedicated_ Bitcoin node and Stacks node for your sBTC Signer.
  - Ensure the nodes are provisioned with the minimum hardware requirements
    described [here][0].
  - Nodes should be _exclusively dedicated_ to serve the sBTC Signer. Avoid
    re-using them to serve other clients as that may negatively affect
    performance (no _mock-signing_, no _Stacks API nodes_).

## Monitor new software releases

> - Stay up-to-date with new releases, patches, and security advisories (e.g.,
>   GitHub, mailing lists, Discord).

Reviewer 4: Suggestion (just tossed this together, but I guess you get the idea I'm going for):
```suggestion
- Stay up-to-date with new releases, patches, and security advisories for all used operating systems, software and packages:
  - https://www.cve.org/ is a great resource for popular software packages.
  - Subscribe to receive security notifications from your vendors.
  - Join relevant messaging channels if applicable (i.e. on Discord, Slack, etc.).
```

Reviewer 2: Running open source vulnerability tools in the CI build are also smart, but I could see an argument that this is additionally something that upstream should aggressively track if it's dependancies in the app.

- Exercise vulnerability management for all packages.
- Apply updates as quickly as possible, especially those addressing a security

>   vulnerability.

Reviewer 4: ```suggestion
  vulnerability.
- Use inventory and patch management software, if available.
```

## Ensure redundancy in operations

> - Ensure that multiple, trusted users can manage and maintain your sBTC Signer instance.
> - Where feasible, users should span different timezones.

Reviewer 4: ```suggestion
- Ensure that multiple, trusted users can manage and maintain your sBTC Signer instance.
- Where feasible, users should span different time zones.
- Document your operations procedures and ensure that relevant personnel have access to them.
```

The following best practices suggest how to create a resilient setup for running
your sBTC Signer.

## Protect your private key and have a cold-storage backup

- Prevent unauthorised access to the sBTC Signer private key.
- Keep an offline, secure backup of your sBTC Signer private key (e.g., hardware
  security modules or encrypted storage devices).

## Backup your sBTC Signer PostgreSQL DB

- Perform daily backups of the sBTC Signer PostgreSQL DB.
- Periodically verify the integrity of backups (e.g. by importing them into a
  fresh PostgreSQL instance).

## Setup proper access control

- Require hardware 2FA keys for access control (e.g., by using Yubikey) to
  connect through SSH, to authenticate to AWS, and for every other relevant
  action.
- Follow the principle of _least privilege_: if you don’t need access, you don’t
  get access; if you get access, it expires after the action is taken.
- _Optional, but strongly recommended_: Implement a "_4-eyes_" process to access
  critical resources (e.g., deploy a new version of the sBTC signer).

## Maintain a strict firewall configuration

- Allow connections to your signer `listen_on` address (used for P2P
  communication).

> - Do not expose any non-essential service to the internet: use a `DEFAULT DENY` policy with explicit `ALLOW`s for necessary network traffic (such as signer p2p and SSH).
>   traffic to _every other port and service_, unless required, e.g. for SSH.

Reviewer 4: Whoops, this got mucked up a little:

```suggestion
- Do not expose any non-essential service to the internet: use a `DEFAULT DENY` policy with explicit `ALLOW`s for necessary network traffic (such as signer p2p and SSH).
```
